File: ts/src/scenarios/coordination-spec.ts
import { z } from "zod";
import { SimulationActionSpecSchema } from "./simulation-spec.js";

export const WorkerSpecSchema = z.object({
  workerId: z.string().min(1),
  role: z.string().min(1),
});

export const CoordinationSpecSchema = z.object({
  description: z.string().min(1),
  environmentDescription: z.string().min(1),
  initialStateDescription: z.string().min(1),
  workers: z.array(WorkerSpecSchema).min(2),
  successCriteria: z.array(z.string().min(1)).min(1),
  failureModes: z.array(z.string().min(1)).default([]),
  actions: z.array(SimulationActionSpecSchema).min(2),
  maxSteps: z.number().int().positive().default(10),
});

export type WorkerSpec = z.infer<typeof WorkerSpecSchema>;
export type CoordinationSpec = z.infer<typeof CoordinationSpecSchema>;

export function parseRawCoordinationSpec(data: Record<string, unknown>): CoordinationSpec {
  return CoordinationSpecSchema.parse({
    description: data.description,
    environmentDescription: data.environment_description,
    initialStateDescription: data.initial_state_description,
    workers: Array.isArray(data.workers)
      ? data.workers.map((worker) => {
          const raw = worker as Record<string, unknown>;
          return {
            workerId: raw.worker_id,
            role: raw.role,
          };
        })
      : data.workers,
    successCriteria: data.success_criteria,
    failureModes: data.failure_modes ?? [],
    actions: data.actions,
    maxSteps: data.max_steps ?? 10,
  });
}

File: autocontext/src/autocontext/scenarios/custom/coordination_spec.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Any

from autocontext.scenarios.custom.simulation_spec import SimulationActionSpecModel


@dataclass(slots=True)
class CoordinationSpec:
    """Spec for a multi-agent coordination scenario."""

    description: str
    environment_description: str
    initial_state_description: str
    workers: list[dict[str, Any]]  # [{worker_id, role, ...}]
    success_criteria: list[str]
    failure_modes: list[str]
    actions: list[SimulationActionSpecModel]
    max_steps: int = 10
